File: backend/Domain/Order.cs
namespace Domain;

public abstract class AuditableEntity
{
    public DateTime CreatedAt { get; set; }
    public DateTime? LastModified { get; set; }
}

public class Order : AuditableEntity
{
    public Guid Id { get; set; }

    public decimal TotalAmount { get; set; }
    public decimal Discount { get; set; }
    public decimal? DeliveryPrice { get; set; }

    public decimal PaidAmount { get; set; }
    public decimal RemainingAmount => Math.Max(0, TotalAmount - PaidAmount);

    public OrderStatus Status { get; set; } = OrderStatus.Ongoing;
    public PaymentStatus PaymentStatus { get; set; } = PaymentStatus.Unpaid;
    public OrderType Type { get; set; }

    public List<OrderItem> OrderItems { get; set; } = new();

    public Guid? TableId { get; set; }
    public Table? Table { get; set; }

    public Guid? CustomerInformationId { get; set; }
    public CustomerInformation? CustomerInformation { get; set; }

    public List<Payment> Payments { get; set; } = new();
}

public enum OrderStatus
{
    Ongoing,
    Completed,
    Closed,
    Cancelled,
}

public enum PaymentStatus
{
    Unpaid,
    PartiallyPaid,
    Paid,
}

public enum OrderType
{
    DineIn,
    Takeaway,
    Delivery,
}

File: frontend/helpers/utils/order.ts
import { ORDER_STATUSES, ORDER_TYPES } from '../constants/constants';
import languagePacks, { LanguageTypes } from '../constants/languagePacks';

export const getOrderStatus = (
    status: ORDER_STATUSES,
    language: LanguageTypes
) => {
    const {
        entities: {
            order: {
                statuses: {
                    ongoing,
                    pendingPayment,
                    cancelled,
                    completed,
                    closed,
                },
            },
        },
    } = languagePacks[language];

    switch (status.toLowerCase()) {
        case ORDER_STATUSES.ONGOING.toLowerCase():
            return ongoing;

        case ORDER_STATUSES.COMPLETED.toLowerCase():
            return completed;

        case ORDER_STATUSES.PENDING_PAYMENT.toLowerCase():
            return pendingPayment;
        case ORDER_STATUSES.CLOSED.toLowerCase():
            return closed;

        case ORDER_STATUSES.CANCELLED.toLowerCase():
            return cancelled;

        default:
            return 'Unknown';
    }
};

export const getOrderType = (type: ORDER_TYPES, language: LanguageTypes) => {
    const {
        entities: {
            order: {
                type: { dinein, takeaway, delivery },
            },
        },
    } = languagePacks[language];

    switch (type.toLowerCase()) {
        case ORDER_TYPES.DINEIN.toLowerCase():
            return dinein;

        case ORDER_TYPES.TAKEAWAY.toLowerCase():
            return takeaway;

        case ORDER_TYPES.DELIVERY.toLowerCase():
            return delivery;

        default:
            return 'Unknown';
    }
};
